sozu: 0.11.56 -> 0.12.0, nixos/sozu: init

netcrns opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
We need Sozu at work, and this makes it more convenient to set up.

###### Things done
Added service configurations for `Sozu` and updated the package's version.
I'm also adding myself to the list of maintainers, to ensure it stays up-to-date.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

netcrns committed:
nixos/sozu: init

nixos/sozu: fix typos to address failed checks

NixOS documentation couldn't build due to typos.
This fixes that.

netcrns committed:
sozu: 0.11.56 -> 0.12.0

SuperSandro2000 commented on pkgs/servers/sozu/default.nix:
```diff
@@ -2,16 +2,16 @@
-  cargoSha256 = "sha256-F5EjBWHBNaQipxCi9Kiz5UTPMU0DuRf15NIVVvhiRxY=";
+  cargoSha256 = "1y5xxldjccdxn31b2wzdy71jh5kdxl3jwslf1haf83jicfvzanba";
 
   buildInputs =
     lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.Security;
```
```suggestion
  buildInputs = lib.optionals stdenv.isDarwin [ Security ];
```

netcrns replied:
that `[ Security ]` makes the build fail, complaining about the variable `Security` not being defined.
importing it gives me this error:
```
error: anonymous function at /home/crns/code/repos/NixOS/nixpkgs/pkgs/servers/sozu/default.nix:1:1 called without required argument 'Security', at /home/crns/code/repos/NixOS/nixpkgs/lib/customisation.nix:69:16
```
what am I missing?

SuperSandro2000 replied:
You are missing https://www.github.com/NixOS/nixpkgs/blob/02a567192e752374c754302be82f9ea8d9b036a2/pkgs/top-level/all-packages.nix#L10579

SuperSandro2000 commented on pkgs/servers/sozu/default.nix:
```diff
@@ -21,6 +21,6 @@ rustPlatform.buildRustPackage rec {
       "Open Source HTTP Reverse Proxy built in Rust for Immutable Infrastructures";
```
This should be in one line.

SuperSandro2000 commented on pkgs/servers/sozu/default.nix:
```diff
@@ -21,6 +21,6 @@ rustPlatform.buildRustPackage rec {
       "Open Source HTTP Reverse Proxy built in Rust for Immutable Infrastructures";
     homepage = "https://www.sozu.io";
     license = licenses.agpl3;
```
agpl3Plus or agpl3Only?

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+    };
+
+    users.users.sozu = {
+      description = "Sozu Daemon User";
```
isSystemUser or isNormalUser needs to be set now.

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+
+    systemd.services.sozu = {
+      description = "Sozu - A HTTP reverse proxy, configurable at runtime, fast and safe, built in Rust.";
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
```
```suggestion
      after = [ "network.target" ];
      wants = [ "network.target" ];
```

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+
+      serviceConfig = {
+        PIDFile = "/run/sozu/sozu.pid";
+        ExecStart = "${pkgs.sozu}/bin/sozu start --config ${configFile}";
```
It would be nice if the package is configurable.

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+    };
+
+    extraConfig = mkOption {
+      default = '''';
```
```suggestion
      default = "";
```

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+          };
+
+          tls_provider = mkOption {
+            type = types.str;
```
Please use an enum with the two types.

netcrns replied:
how do you mean?

SuperSandro2000 replied:
something like 
https://github.com/nix-community/home-manager/blob/db00b39a9abec04245486a01b236b8d9734c9ad0/modules/accounts/email.nix#L191

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+            default = 2;
+            description = ''
+              The number of worker processes that will handle traffic.
+              Defaults to 2.
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+              sent to sozu is too large, or the data that sozu must return is too large, the
+              buffer will grow up to max_command_buffer_size. If the buffer is still not large
+              enough sozu, will close the connection.
+              Defaults to 1000000.
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,299 @@
+        options = {
+          saved_state = mkOption {
+            default = "/run/sozu/state.json";
+            type = types.str;
```
```suggestion
            type = types.file;
```

netcrns replied:
`types.file` doesn't exist, and `types.path`is not what I want here.
`types.str` has been choses intentionally.

netcrns replied:
Is this still a concern I am missing the point of, or can I resolve this?

aanderse replied:
I'll be addressing this point later anyways, so you can ignore for now.

SuperSandro2000 reviewed (commented):
Modules do not have a version. Naming the commit ``nixos/sozu: init`` is fine. Also please squash the 2nd module commit.

search.nixos.org shows the defaults and examples by default. They can be removed.

and the module is missing a maintainer.

aanderse commented:
Sorry this PR isn't a quick one to review and I haven't found the time yet. I'll start with one comment: with modern software like this you would think upstream would suggest running as a non root user... but I couldn't find any mention on their github project. Would you mind asking the question to upstream on what it would take to run the software as a non root user, taking advantage of systemd to close the gap?

I'll try to find the time this PR deserves for review soon.

Br1ght0ne commented:
I'll wait with my review until other questions are resolved.

Geal commented:
hi, I'm the lead developer for sozu. I just added some documentation about making privileged ports accessible as a non root user. I thought this was common by now, but when looking for clear docs, there was a lot of confusing information. I hope this helps: https://github.com/sozu-proxy/sozu/blob/master/doc/recipes.md#using-port-80-or-443-as-non-root-user

Let me know if there are some things you need to configure the package (FYI sozu 0.12 will be released very soon)

aanderse commented:
@Geal thank you for a clear and concise answer. Much appreciated!

netcrns commented:
Squashed stuff to bring some order into my mess.
I've also removed myself from the maintainers-list, in favour of doing that [here](https://github.com/NixOS/nixpkgs/pull/120754), seeing as that PR will probably be merged before this one.

SuperSandro2000 commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,304 @@
+  };
+
+  meta = with lib; {
+    maintainers = with maintainer; [ netcrns ];
```
Please add yourself as a maintainer by adding an alphabetically sorted entry to `maintainers/maintainer-list.nix` and commit this change into a separate commit with the message ``maintainers: add NAME``.

netcrns replied:
> https://github.com/NixOS/nixpkgs/pull/120010#issuecomment-827008692
I've also removed myself from the maintainers-list, in favour of doing that [here](https://github.com/NixOS/nixpkgs/pull/120754), seeing as that PR will probably be merged before this one.

> https://github.com/NixOS/nixpkgs/pull/120754#issuecomment-826974445
Seeing as this PR should presumably be merged before the aforementioned one (we're waiting for sozu v0.12.0 to be released), I'll remove the commit from that and move it here. Would that be alright?

netcrns commented:
I'm not sure how I'll be able to make privileged ports accessible to non-root users, as that would require making the `sozu` user controlling the server unprivileged, if I'm not mistaken. Having it unprivileged makes it unable to create the `/run` directory, and I'm unsure as to how to create it if not in `systemd.services.sozu.preStart`.

netcrns committed:
nixos/sozu: make privileged ports accessible to non-root users

lukegb commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,303 @@
+  cfg = config.services.sozu;
+  format = pkgs.formats.toml {};
+
+  settings = format.generate "settings.toml" cfg.settings;
+
+  configFile = pkgs.writeText "config.toml" ''
+    ${fileContents settings}
+
+    ${optionalString (cfg.extraConfig != null) cfg.extraConfig}
+  '';
```
Hi! Thanks for doing this, but community consensus is that, in most cases, we shouldn't attempt to describe config options unless doing so is required to make other parts of the module work.

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

netcrns replied:
I tried my best to avoid doing what I did in the end, but couldn't manage to completely stick with the structural `settings` pattern. As this took me much longer than expected, I decided to — before wasting even more time — just create the PR and see what people say.

The main issue I had was making options _completely_ optional, which I was told should be done with `types.nullOr types.<expected type>`. This was impossible due to the apparent nature of the TOML formatter which doesn't accept null-values, but also seems to need a declared value for every defined option (either a default value or one set by the user, but it _has_ to be there).
I think there were some more issues that I can't accurately remember at the moment; they'll probably come up eventually.

There most likely is a very simple way to do this that I just couldn't figure out.
I did try asking around on Discord, but didn't really manage to get an answer.

lukegb replied:
Hmm. In general, most of Sozu's options shouldn't be declared in the NixOS config, so this shouldn't be a problem? I might be misunderstanding but the entire settings option should be typed as a `types.submodule { freeformType = format.type; };` and left at that.

netcrns replied:
Oh, wow... I've tried this before, but completely misunderstood the purpose of `freeformType`; this makes so much more sense.
I'll get right back to making things better in a bit. Big thanks for the clarification!

netcrns replied:
Right, so, this works wonderfully - but I'm not sure how to make the documentation for what's not declared in `sozu.nix` visible & clear to the user.
How do you think should I approach this?

For example:
```nix
{
  services.sozu = {
    # Has description
    enable = true;

    # Has description
    settings = {

      # Does not have description, since it's not a declared/strictly required option
      metrics = {
        address = "<whatever>";
      };

      # Does not have description, since it's not a declared/strictly required option
      listeners = [
        { protocol = "http"; } {protocol = "https"; } 
      ];
    };
  };
}
```
This works, and is needed to actually set listeners and yada yada up, but that's not at all clear to the user of this service.
Where should I describe this process, if it's freeform?

netcrns replied:
Yes, but the toml documentation does in no way suffice for a new nix user. I can say this from experience; I just know how to do it because I wrote this module and I had to look into it - which is why I feel obligated to make the process easier for those using said module. Isn't that what this is ultimately supposed to do?

aanderse replied:
You should write an abstraction in the form of an `option` instead of having an `extraConfig` option. This situation is (yet another way) where NixOS can shine and add so much value to the sysadmin.

lukegb commented:
Sozu looks super neat! I'm gonna add it to my pile of things to look at, heh.

netcrns committed:
nixos/sozu: address comments in PR

nixos/sozu: remove `extraConfig` in favour of `optional`

nixos/sozu: rename `settings` to `global`

nixos/sozu: fix defaults

nixos/sozu: add examples to freeform submodule

nixos/sozu: fix environment package to be value of `package` option

nixos/sozu: assign group to user `sozu`

nixos/sozu: stylistic changes

netcrns commented:
/marvin opt-in
/status needs_reviewer

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,303 @@
+        ExecStart = "${cfg.package}/bin/sozu start --config ${configFile}";
+        ExecReload = "${cfg.package}/bin/sozuctl --config ${configFile} reload";
+        Restart = "on-failure";
+        User = "sozu";
```
Please create a corresponding group, or run as `DynamicUser`.

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,303 @@
+      isSystemUser = true;
+    };
+
+    systemd.services.sozu = {
```
Given the advertised reload ability with zero downtime that `sozu` boasts on its github page this unit seems tragic... **any** little change will cause this service to fully restart.

netcrns replied:
Oh, I'm honestly not sure what I could improve here - I just copied & adjusted the unit provided by upstream.

aanderse replied:
Have you validated that the `sozuctl ... reload` command works? I can't seem to have any success with it.

Here is my machine configuration:
```nix
  services.httpd.enable = true;
  services.httpd.adminAddr = "webmaster@example.org";

  services.httpd.virtualHosts.site0 = {
    listen = [ { ip = "0.0.0.0"; port = 8080; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #0</html>";
  };

  services.httpd.virtualHosts.site1 = {
    listen = [ { ip = "0.0.0.0"; port = 8081; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #1</html>";
  };

  services.httpd.virtualHosts.site2 = {
    listen = [ { ip = "0.0.0.0"; port = 8082; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #2</html>";
  };

  services.httpd.virtualHosts.site3 = {
    listen = [ { ip = "0.0.0.0"; port = 8083; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #3</html>";
  };

  services.sozu.enable = true;
  services.sozu.settings.applications."applications.localhost" = {
    protocol = "http";
    frontends = [ { address = "0.0.0.0:80"; hostname = "localhost"; } ];
    backends = [
      { address = "127.0.0.1:8080"; }
      # toggle a few of these and hit `curl localhost`
      # { address = "127.0.0.1:8081"; }
      # { address = "127.0.0.1:8082"; }
      { address = "127.0.0.1:8083"; }
    ];
  };
```

Additionally, I applied this patch on top of your PR:
```patch
diff --git a/nixos/modules/services/networking/sozu.nix b/nixos/modules/services/networking/sozu.nix
index 75be7642644..9589677eea0 100644
--- a/nixos/modules/services/networking/sozu.nix
+++ b/nixos/modules/services/networking/sozu.nix
@@ -31,6 +31,7 @@ in
         Documentation <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md">here</link>.
       '';
       type = types.submodule {
+        freeformType = format.type;
 
         options = {
           saved_state = mkOption {
@@ -268,29 +269,22 @@ in
   };
 
   config = mkIf cfg.enable {
-    environment.systemPackages = with pkgs; [
-      sozu
-    ];
-
-    users.users.sozu = {
-      description = "Sozu Daemon User";
-      isSystemUser = true;
-    };
-
     systemd.services.sozu = {
       description = "Sozu - A HTTP reverse proxy, configurable at runtime, fast and safe, built in Rust.";
       after = [ "network.target" ];
       wants = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
 
+      reloadIfChanged = true;
+
       serviceConfig = {
+        DynamicUser = true;
         PIDFile = "/run/sozu/sozu.pid";
         ExecStart = "${cfg.package}/bin/sozu start --config ${configFile}";
         ExecReload = "${cfg.package}/bin/sozuctl --config ${configFile} reload";
         Restart = "on-failure";
-        User = "sozu";
         RuntimeDirectory = "sozu";
-        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
       };
 
     };
```

**NOTE:** the only relevant parts of this patch are `systemd.services.sozu.reloadIfChanged = true;` and the addition of `freeformType = format.type;` to make my testing easier.

When I toggle some of the backends on and off, then `reload`, I still hit all 4 backends. Any ideas? I've never used `sozu` before, so maybe I'm missing something fundamental...

aanderse replied:
I realized this is likely the expected behavior because you're not passing a static path to the configuration file. Actually the same thing exists with the `httpd` module currently... I should fix that.

aanderse replied:
I fixed this issue in #122658. You should do the same thing here. I think we can get zero downtime updates with this software on NixOS.

aanderse reviewed (commented):
I think some of my comments might be in the realm of "advanced" instead of "beginner". I have no idea what your experience and comfort level with nixos is, so please do not hesitate to ask for help if you need. Or ignore some of my comments. 🤷‍♂️

netcrns committed:
nixos/sozu: add description to `optional`

netcrns committed:
nixos/sozu: provide a static path to the config file

netcrns committed:
nixos/sozu: take strings and paths, instead of just strings

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,479 @@
+      description = "Sozu package to use.";
+    };
+
+    configFile = mkOption {
```
Are there configurations which aren't possible with `settings`, or is this for convenience to hand write your own config?

netcrns replied:
It's for convenience.

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,479 @@
+      '';
+    };
+
+    global = mkOption {
```
You should have a single option called `settings` instead of `global` and `optional`.

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,479 @@
+      '';
+      type = types.submodule {
+
+        options = {
```
Are all of these options required? It seems like some people might **not** want to include some of these options into their configuration from the description. I would suggest you go for a more minimal approach here (less typed options) and fill in values in the `config` if needed.

netcrns replied:
Let me know if it's good now.

aanderse replied:
I was thinking something a little bit more simple like [this](https://github.com/aanderse/nixpkgs/blob/cfad31a7310d1b3d1723f53598fd65f3a069bfd9/nixos/modules/services/networking/sozu.nix).

Note that you could still using a `freeformType` to provide restrictions around some options if needed, though it is likely just overhead that doesn't provide as much value in this case... most of the general configuration options are pretty self explanatory.

aanderse replied:
Also worth mentioning is the removal of the `configFile` option. This complicates things because `restartTriggers` stops working and the module becomes pretty much worthless. You can't reference anything out of `settings` anymore because it might be wrong.

I think the solution to having users use `toml` natively is `builtins.fromToml`, perhaps. Maybe `services.sozu.settings = builtins.fromToml /path/to/my/config.toml;`.

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,479 @@
+        restartTriggers = [ cfg.configFile ];
+
+        serviceConfig = {
+          PIDFile = "/run/sozu/sozu.pid";
```
Setting a hard coded value here works against the above options.

aanderse commented on nixos/modules/services/networking/sozu.nix:
```diff
@@ -0,0 +1,479 @@
+          }
+        ];
+
+        applications = {
```
I'm not really sure about this, especially since I've never used `sozu`, but I feel like there might be value in creating a NixOS option for `applications`. Maybe you benefit from some automatic certificate integration with our `security.acme` module, maybe it makes it easier to provide examples... I'm not sure yet. I'll give this some more thought.

Having a single `settings` option might feel a bit overwhelming to do entirely in `nix` instead of `toml` for some people, if they're coming from a non NixOS distro as an existing `sozu` user.

netcrns replied:
I'll probably need help with this, as I don't know how to go about integrating `security.acme`.

aanderse reviewed (commented):
I've left a few more suggestions. Please let me know if you need a hand.

netcrns committed:
nixos/sozu: address comments in pr

nixos/sozu: make `PIDFile` take `cfg.settings.pid_file_path`

nixos/sozu: move options previously in `global` into top-level

nixos/sozu: rename `optional` to `settings`

nixos/sozu: move potentially unwanted options into `settings`

nixos/sozu: add options to `settings`
  * listeners
  * applications
  * metrics

netcrns committed:
nixos/sozu: add descriptions to options that are missing them

netcrns committed:
sozu: use correct cargo hash

Not sure why this changed in the first place, but there was a conflict
with `master`.

netcrns commented:
shush marvin
/status awaiting_changes

timokau commented:
Hi!

The marvin-mk2 bot is now [discontinued](https://github.com/timokau/marvin-mk2/issues/34#issuecomment-1100656280). I have removed the relevant tags from this PR. If you still need someone to look at it, one option would be to ask in [this discourse thread](https://discourse.nixos.org/t/prs-ready-for-review/3032).

I am posting this notice to all open PRs with the `marvin` tag. Please understand that following all of these discussions would take too much time, so I will unsubscribe from this PR unless I have already been involved in it before this message.

wegank commented:
sozu is at 0.15.6 on master, please rebase and reopen a new PR with updated options.
